src/libraries/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs

```diff
@@ -266,11 +266,16 @@ internal void MoveRule(int i)
             if (i > 1)
             {
                 T[] rules = Rules!;
-                T rule = rules[i];
+                // Synchronization of AddRule is omitted for performance. Concurrent invocations of AddRule
+                // may cause Rules to revert back to an older (smaller) version, making i out of bounds.
+                if (unchecked((uint)i < (uint)rules.Length))
```

Comment:
Please consider using `if (i < rules.Length)`.

Reply:
Simplified to `if (i < rules.Length)`. Thanks!